src/cloud/lib/editor/CodeMirror.ts

```diff
@@ -21,3 +22,239 @@ export interface EditorPosition {
+    { text: 'sql', displayText: 'SQL' },
+    { text: 'stex', displayText: 'sTeX, LaTeX' },
+    { text: 'stylus', displayText: 'Stylus' },
+    { text: 'swift', displayText: 'Swift' },
+  ],
+  t: [
+    { text: 'tcl', displayText: 'Tcl' },
+    { text: 'textile', displayText: 'Textile' },
+    { text: 'tiddlywiki', displayText: 'Tiddlywiki' },
+    { text: 'tiki', displayText: 'Tiki wiki' },
+    { text: 'toml', displayText: 'TOML' },
+    { text: 'tornado', displayText: 'Tornado (templating language)' },
+    { text: 'troff', displayText: 'troff (for manpages)' },
+    { text: 'ttcn', displayText: 'TTCN' },
+    { text: 'ttcn-cfg', displayText: 'TTCN Configuration' },
+    { text: 'turtle', displayText: 'Turtle' },
+    { text: 'twig', displayText: 'Twig' },
+  ],
+  v: [
+    { text: 'vb', displayText: 'VB.NET' },
+    { text: 'vbscript', displayText: 'VBScript' },
+    { text: 'velocity', displayText: 'Velocity' },
+    { text: 'verilog', displayText: 'Verilog/SystemVerilog' },
+    { text: 'vhdl', displayText: 'VHDL' },
+    { text: 'vue', displayText: 'Vue.js app' },
+  ],
+  w: [
+    { text: 'webidl', displayText: 'Web IDL' },
+    { text: 'wast', displayText: 'WebAssembly Text Format' },
+  ],
+  x: [
+    { text: 'xml', displayText: 'XML/HTML' },
+    { text: 'xquery', displayText: 'XQuery' },
+  ],
+  y: [
+    { text: 'yacas', displayText: 'Yacas' },
+    { text: 'yaml', displayText: 'YAML' },
+    { text: 'yaml-frontmatter', displayText: 'YAML frontmatter' },
+  ],
+  z: [{ text: 'z80', displayText: 'Z80' }],
+}
+
+export function getModeSuggestions(
+  word: string,
+  suggestionModes: {
+    [key: string]: { text: string; displayText: string }[]
+  } = supportedModeSuggestions
+) {
+  for (const [key, suggestions] of Object.entries(suggestionModes)) {
+    if (word.startsWith(key)) {
+      const filteredSuggestions = suggestions.filter((suggestion) =>
+        suggestion.text.startsWith(word)
+      )
+      // if user accepted suggestion - don't show suggestion which auto-completes to the same string
+      // if (
+      //   filteredSuggestions.length == 1 &&
+      //   word == filteredSuggestions[0].text
+      // ) {
+      //   return []
+      // }
+      return filteredSuggestions
+    }
+  }
+  return []
+}
+
+export function CodeMirrorEditorModeHints(cm: CodeMirror.Editor) {
+  return new Promise(function (accept) {
+    setTimeout(function () {
+      const cursor = cm.getCursor(),
+        line = cm.getLine(cursor.line)
+      let start = cursor.ch
+      let end = cursor.ch
+      while (start && /\w/.test(line.charAt(start - 1))) --start
+      while (end < line.length && /\w/.test(line.charAt(end))) ++end
+      const word = line.slice(start, end).toLowerCase()
+      const suggestions = getModeSuggestions(word)
+      if (suggestions.length == 0) {
+        return accept(null)
+      }
+      // todo: [komediruzecki-2022-03-27] no suggestions if we already have matched
```

Comment:
As I said above, keep showing suggestion until a user actually choose one from the list.

Reply:
This is tricky, not sure how to save such state, and when to actually allow opening it again (on another editor line etc.)